server.py

```diff
@@ -23,53 +28,137 @@
-# Endpoint for PUT method
-@app.route('/write', methods=['PUT'])
+@app.route('/redirect')
+def redirect_endpoint():
+    return redirect('http://localhost:5000/read?uuid=xxx', code=302)
+
+# Endpoint for write method
+@app.route('/write', methods=['POST'])
 def write_file():
     file = request.files['file']
     if file:
         filename = secure_filename(file.filename)
         file_uuid = str(uuid.uuid4())
-        # return str(os.path.join(app.config['UPLOAD_FOLDER'], file_uuid)), 200
         file.save(os.path.join(app.config['UPLOAD_FOLDER'], file_uuid))
+        metadata = metadata_manager.MetadataManager()
+        metadata.update_file_stored(file_uuid, app.config['HOST'])
+        metadata.close_connection()
         return file_uuid, 201
     return 'Write Failed', 500
 
-# Endpoint for GET method
+# Endpoint for read method
 @app.route('/read', methods=['GET'])
 def read_file():
-    sys.stdout.flush()
     filename = request.args.get('uuid')
-    return send_from_directory(UPLOAD_FOLDER, secure_filename(filename))
+    file_path = UPLOAD_FOLDER + '/' + secure_filename(filename)
+    metadata = metadata_manager.MetadataManager()
+    if (os.path.exists(file_path)):
+        return send_from_directory(UPLOAD_FOLDER, secure_filename(filename))
+
+    redirect_address = metadata.lookup_file(filename, app.config['HOST'])
+    if (redirect_address is not None):
+        url = 'http://%s/read?%s' % (redirect_address, urllib.urlencode({ 'uuid': filename }))
+        return redirect(url, code=302)
+
+    other_servers = metadata.get_all_server(app.config['HOST'])
+    if (len(other_servers) > 0):
+        for server in other_servers:
+            url = 'http://%s/file_exists?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+            lookup_request = requests.get(url)
+            if (lookup_request.status_code == 200):
+                redirection_url = 'http://%s/read?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+                metadata.update_file_stored(filename, server)
+                metadata.close_connection()
+                return redirect(redirection_url, code=302)
+
+    return 'File Not Found', 404
+
+@app.route('/file_exists', methods=['GET'])
+def file_exists():
+    filename = request.args.get('uuid')
+    file_path = UPLOAD_FOLDER + secure_filename(filename)
+    print file_path
+    if (os.path.exists(file_path)):
+        return app.config['HOST'], 200
+    else:
+        return 'File not found', 404
 
+# Helper method for sending a file to another server
 def move_file(request):
     file_uuid = request.args.get('uuid')
     destination = request.args.get('destination')
     destination_with_endpoint = destination + '/write'
-    write_request = util.construct_put_request(destination_with_endpoint, file_uuid)
+    write_request = util.construct_post_request(destination_with_endpoint, file_uuid)
+    metadata = metadata_manager.MetadataManager()
+    metadata.update_file_stored(file_uuid, destination)
```

Comment:
shouldn't we wait to update this until the request returns? otherwise we might have some hanging metadata with no way to validate/delete it.